src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py

```diff
@@ -131,32 +131,36 @@ def handle(self, *args, **options):
         if options["uuid"]:
             package_uuid = options["uuid"]
             self.info("Reindexing package {}".format(package_uuid))
-            package_info = storageService.get_file_info(uuid=package_uuid)
+            try:
+                package_info = storageService.get_file_info(uuid=package_uuid)[0]
+            except IndexError:
+                self.error(
+                    "Error indexing package {0}. "
+                    "Package with supplied UUID not found in Storage Service".format(
+                        package_uuid
+                    )
+                )
+                return
 
             is_aic = False
             package_type = package_info.get("package_type")
             if package_type and package_type == "AIC":
                 is_aic = True
 
-            (index_success, error_message) = self.process_package(
+            index_success = self.process_package(
                 es_client,
                 package_info,
                 temp_dir,
                 delete_before_reindexing,
                 is_aic=is_aic,
             )
-            if index_success is True:
-                self.info("Successfully reindexed package {}".format(package_uuid))
-            else:
-                self.error(
-                    "Error indexing package {0}. Details: {1}".format(
-                        package_uuid, error_message
-                    )
-                )
 
             self.info("Cleaning up")
             shutil.rmtree(temp_dir)
-            self.info("Indexing complete.")
+            if index_success:
+                self.success("Success! Indexing complete.")
+            else:
+                self.error("Failure. Indexing complete.")
             return
```

Comment:
We can remove this `return` too.

Reply:
See above.

Reply:
Ordering went off with the refactor, so that comment might not be so helpful. I believe this return is necessary to prevent the task from continuing to index other packages when the --uuid flag is used.

Reply:
That makes sense. I feel like when there's a return like this in the function it says there might be something not quite right with the flow-control. So, in the code above we're directing the flow if `uuid` is set. Given that, I'd almost be tempted to say:
```python
if options["uuid"]:
   ...do thing
else:
   ...do the other thing
```
But that doesn't quite feel right either. I suppose a `return` does keep everything to the left quite nicely. 

But then I thought to myself what else might we be missing.

I had one observation, and that was the setup for Elasticsearch came too early. What happens if we retrieve zero packages from the storage service? e.g. there's a connection error? I think we will have deleted our index and didn't need to.

The second observation I had with the new function for indexing a single package is that we really wanted to do the same thing across both main sections of the code. Given the storage service helper always returns a list, we can treat a list of 1 (or zero) the same we treat a list of 100, with a few small exceptions written along the way.

I have modified your functions here. So I have removed `index_single_package` and moved the bulk of the unique work inline so that most of the logic is the same for bulk or single operations. Also, we now only setup Elasticsearch just in time, so that we only do anything with the index if there is something to do. See what you think:

```diff
diff --git a/src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py b/src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py
index f766ee77..d48c05fa 100644
--- a/src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py
+++ b/src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py
@@ -116,9 +116,6 @@ class Command(DashboardCommand):
         logging.getLogger("elasticsearch").setLevel(logging.ERROR)
         logging.getLogger("archivematica.common").setLevel(logging.ERROR)
 
-        # Setup es_client and delete indices if required.
-        es_client = setup_es_for_aip_reindexing(self, options["delete_all"])
-
         # Create temporary directory for downloaded METS files.
         temp_dir = tempfile.mkdtemp()
 
@@ -129,29 +126,27 @@ class Command(DashboardCommand):
         if options["delete"]:
             delete_before_reindexing = True
 
-        # Index a single package and return if uuid option is enabled.
-        if options["uuid"]:
-            index_success = self.index_single_package(
-                options["uuid"], es_client, temp_dir, delete_before_reindexing
-            )
-            if index_success:
-                self.success("Success! Indexing complete.")
-            return
-
         self.info("Rebuilding 'aips' and 'aipfiles' indices")
-
         packages_not_indexed = []
-
-        # Index AIPs and AICs that are associated with this pipeline
-        # that are not deleted or replicas.
-        packages = storageService.get_file_info()
-        aips_to_index = storageService.filter_packages(
-            packages,
-            package_types=PACKAGE_TYPES_TO_INDEX,
-            pipeline_uuid=pipeline_uuid,
-            filter_replicas=True,
-        )
+        if not options["uuid"]:
+            # Index AIPs and AICs that are associated with this pipeline
+            # that are not deleted or replicas.
+            packages = storageService.get_file_info()
+            aips_to_index = storageService.filter_packages(
+                packages,
+                package_types=PACKAGE_TYPES_TO_INDEX,
+                pipeline_uuid=pipeline_uuid,
+                filter_replicas=True,
+            )
+        else:
+            aips_to_index = storageService.get_file_info(uuid=options["uuid"])
         aips_to_index_count = len(aips_to_index)
+        if aips_to_index_count > 1:
+            # For bulk operations setup es_client and delete indices if required.
+            es_client = setup_es_for_aip_reindexing(self, options["delete_all"])
+        elif aips_to_index_count:
+            # For single item operations, setup es_client and don't delete indices.
+            es_client = setup_es_for_aip_reindexing(self, False)
         aip_indexed_count = 0
         for aip in aips_to_index:
             is_aic = False
@@ -184,37 +179,6 @@ class Command(DashboardCommand):
                 )
             )
 
-    def index_single_package(self, uuid, es_client, temp_dir, delete_before_reindexing):
-        """Index single package by UUID.
-
-        :param uuid: Package UUID.
-        :param es_client: Elasticsearch client.
-        :param temp_dir: Path to tempdir for downloaded METS files.
-        :param delete_before_reindexing: Boolean of whether to delete
-        package from indices prior to reindexing.
-
-        :returns: Boolean indicating success.
-        """
-        self.info("Reindexing package {}".format(uuid))
-        try:
-            package_info = storageService.get_file_info(uuid=uuid)[0]
-        except IndexError:
-            self.error("Error indexing package {0}. UUID not found.".format(uuid))
-            return False
-
-        is_aic = False
-        package_type = package_info.get("package_type")
-        if package_type and package_type == "AIC":
-            is_aic = True
-
-        index_success = self.process_package(
-            es_client, package_info, temp_dir, delete_before_reindexing, is_aic=is_aic
-        )
-
-        self.info("Cleaning up")
-        shutil.rmtree(temp_dir)
-        return index_success
-
     def process_package(
         self, es_client, package_info, temp_dir, delete_before_reindexing, is_aic=False
     ):
```
You might see other optimizations, but does that help at all? (I wasn't quite sure how to write this down in a better narrative, so hoped maybe the code-diff would work as well).

Reply:
It did help! I refactored based on your comments here (FWIW, the diff really did help me come around to your point) and the resulting code is shorter, more readable, and handles the case of no packages being returned by the Storage Service much more gracefully.